# 1 - Define the problem / Questions

# 2. Data Collections -

# 3 - Data Cleaning  -
- Handle missing values

- Identify and remove duplicates

- handle outliers

- Transform the data into suitable format

# 4- EDA - Exploratory Data Analysis.

# 5- Data Visualization -

# 6 - Analyze the insights & Report


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
titanic = sns.load_dataset('titanic')

In [3]:
#initial data inspection

In [4]:
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [5]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [6]:
#As we seen , the data contain some missing values.
#so lets check the exact number

In [8]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

#insights

- age - 177 missing values - should be imputed based on pclass or sex - replace by mean or median

- embarked - 2 missing values - can be filled with mode

- deck - massive missing values - better to drop this column


# .fillna()

# .dropna - .dop()

In [9]:
#fill with median age based on pclass & sex

In [11]:
titanic['age'] =titanic.groupby(['pclass', 'sex'])['age'].transform(lambda x : x.fillna(x.median()))

In [12]:
titanic['age'].isnull().sum()

0

In [13]:
#replace embark with mode

In [26]:
titanic['embarked'].mode()[0]

'S'

In [29]:
titanic['embarked'].fillna(titanic['embarked'].mode()[0], inplace = True)
#inplace = True is for permenent changes

In [30]:
titanic['embarked'].isnull().sum()

0

In [31]:
titanic.drop(columns=['deck'], inplace= True)

In [32]:
titanic.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [33]:
titanic.drop(columns=['who', 'embark_town', 'alive', 'adult_male'] , inplace= True)

In [34]:
titanic.isnull().sum()

survived    0
pclass      0
sex         0
age         0
sibsp       0
parch       0
fare        0
embarked    0
class       0
alone       0
dtype: int64

In [35]:
titanic.shape

(891, 10)

After dropping unwanted columns the titanic dataset now has 891 rows describe the 891 passenger details and 10 columns which shows the passenger features.

In [36]:
#lets check for statistical summery of the data -

In [47]:
titanic.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   773.000000  773.000000   773  773.000000  773.000000  773.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   482         NaN         NaN         NaN   
mean      0.415265    2.247089   NaN   29.558111    0.529107    0.421734   
std       0.493087    0.853070   NaN   13.988257    0.991280    0.841380   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    1.000000   NaN   21.000000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    1.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class alone  
count   773.000000      773    773   773  
unique         NaN        3      3     2  
top            NaN        S  Third  True  
freq           NaN      562    400   436  
mean     35.003315      NaN    NaN   NaN  
std      52.443053      NaN    NaN   NaN  
min       0.000000      NaN    NaN   NaN  
25%       8.050000      NaN    NaN   NaN  
50%      16.100000      NaN    NaN   NaN  
75%      34.375000      NaN    NaN   NaN  
max     512.329200      NaN    NaN   NaN

 - survival rate only 38% while 62% did not survived

- half of the passenger were in third class

- most passenger were young adult - median age = 26 year

- must be outlier in age shown 80 year old also

- median fare 14.45 but max was 512.33 - it also may contain outliers

-most passenger travel alone

- most passenger embarked from 'S'

- most passenger class is Third class

In [41]:
titanic.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
886     True
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

In [42]:
titanic.duplicated().sum()

118

In [45]:
titanic.drop_duplicates(inplace= True)

In [46]:
titanic.duplicated().sum()

0